### PR TITLE
node: cpumgr: fix allocation of full physical CPUs when the policy option `full-pcpus-only` is enabled

### LINKED ISSUE #113537: The static cpumanager flag `full-pcpus-only` not ensure to allocate full physical cores
### What happened?

I run a unit test in `pkg/kubelet/cm/cpumanager`,  it has 12 vcpu and reserves cpus `1,6`. When the container request 10 vcpu and the FullPCPUsOnlyOption is true, the allocation result is `0,2-5,7-11`. in my opinion, the `0` and `7` vcpu should be ignore because they are single thread.
```
func Test(t *testing.T) {
	topoDualSocketHT := &topology.CPUTopology{
		NumCPUs:    12,
		NumSockets: 2,
		NumCores:   6,
		CPUDetails: map[int]topology.CPUInfo{
			0:  {CoreID: 0, SocketID: 0, NUMANodeID: 0},
			1:  {CoreID: 1, SocketID: 1, NUMANodeID: 1},
			2:  {CoreID: 2, SocketID: 0, NUMANodeID: 0},
			3:  {CoreID: 3, SocketID: 1, NUMANodeID: 1},
			4:  {CoreID: 4, SocketID: 0, NUMANodeID: 0},
			5:  {CoreID: 5, SocketID: 1, NUMANodeID: 1},
			6:  {CoreID: 0, SocketID: 0, NUMANodeID: 0},
			7:  {CoreID: 1, SocketID: 1, NUMANodeID: 1},
			8:  {CoreID: 2, SocketID: 0, NUMANodeID: 0},
			9:  {CoreID: 3, SocketID: 1, NUMANodeID: 1},
			10: {CoreID: 4, SocketID: 0, NUMANodeID: 0},
			11: {CoreID: 5, SocketID: 1, NUMANodeID: 1},
		},
	}
	st := &mockState{
		assignments:   state.ContainerCPUAssignments{},
		defaultCPUSet: cpuset.NewCPUSet(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11),
	}
	policy, _ := NewStaticPolicy(topoDualSocketHT, 2, cpuset.NewCPUSet(1, 6), topologymanager.NewFakeManager(), map[string]string{FullPCPUsOnlyOption: "true"})
	pod := makePod("fakePod", "fakeContainer3", "10000m", "10000m")
	container := &pod.Spec.Containers[0]
	err := policy.Allocate(st, pod, container)
	if err != nil {
		t.Errorf("failed to alloca:%v", err)
	}
	cset, found := st.assignments[string(pod.UID)][container.Name]
	if found {
		t.Logf("allocate:%v", cset)
	}
}
```

output
```
I1102 16:18:04.353881   14825 policy_static.go:356] "AllocateCPUs" result="0,2-5,7-11"
```

### What did you expect to happen?

When we set `full-pcpus-only` with   static cpumanager,  kubelet should ensure always allocate full physical cores, and ignore single vcpu thread.

Like above unit test, the allocation should failed.

### How can we reproduce it (as minimally and precisely as possible)?

Run the unit test in package `pkg/kubelet/cm/cpumanager`

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
# v1.24.6
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by ffromani**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
Make sure we allocate full cores when the `full-pcpus-only`  option is enabled, the existing pre-check is not sufficient.
We need to make sure we never get as far as allocating single CPUs, so the smallest unit we allocate is a virtual thread group.

#### Which issue(s) this PR fixes:
Fixes #113537 

#### Does this PR introduce a user-facing change?
```release-note
Fix misallocation of CPUs when using the `full-pcpus-only` CPUManager option
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
NONE
```

### CODE REVIEW

File: pkg/kubelet/cm/cpumanager/cpu_assignment.go [hunk 1]
```diff
@@ -477,6 +477,13 @@ func takeByTopologyNUMAPacked(topo *topology.CPUTopology, availableCPUs cpuset.C
 		return acc.result, nil
 	}
 
+	if cpuGroupSize > 1 {
+		return cpuset.NewCPUSet(), SMTAlignmentError{
+			RequestedCPUs: numCPUs,
+			CpusPerCore:   topo.CPUsPerCore(),
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply pkg/kubelet/cm/cpumanager/cpu_assignment.go:486]: I'm still looking for the best approach. I'm not too fond of the current changes myself. And yes, we will need the check in both variants.

**Comment by fromanirh** [comment]
/hold
In addition to the comments (which I fully agree with, the current changes could use some serious refinement), I want to make sure the test coverage is better. That was the real culprit back in time, and I want to improve in this area

File: pkg/kubelet/cm/cpumanager/policy_static.go [hunk 2]
```diff
@@ -416,15 +421,11 @@ func (p *staticPolicy) podGuaranteedCPUs(pod *v1.Pod) int {
 	return requestedByAppContainers
 }
 
-func (p *staticPolicy) takeByTopology(availableCPUs cpuset.CPUSet, numCPUs int) (cpuset.CPUSet, error) {
+func (p *staticPolicy) takeByTopology(availableCPUs cpuset.CPUSet, numCPUs, cpuGroupSize int) (cpuset.CPUSet, error) {
 	if p.options.DistributeCPUsAcrossNUMA {
-		cpuGroupSize := 1
-		if p.options.FullPhysicalCPUsOnly {
-			cpuGroupSize = p.topology.CPUsPerCore()
-		}
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply pkg/kubelet/cm/cpumanager/policy_static.go:424]: answered in https://github.com/kubernetes/kubernetes/pull/113924#discussion_r1024209933

File: pkg/kubelet/cm/cpumanager/policy_static.go [hunk 3]
```diff
@@ -343,6 +343,11 @@ func (p *staticPolicy) allocateCPUs(s state.State, numCPUs int, numaAffinity bit
 
 	allocatableCPUs := p.GetAvailableCPUs(s).Union(reusableCPUs)
 
+	cpuGroupSize := 1
+	if p.options.FullPhysicalCPUsOnly {
+		cpuGroupSize = p.topology.CPUsPerCore()
+	}
+
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply pkg/kubelet/cm/cpumanager/policy_static.go:350]: OTOH, honoring the flag also when reserving CPUs at policy creation time will also prevent people to reserve just 1 cpu automatically. Which is probably still legit, but also likely too big/surprising of a change

**Comment by fromanirh** [comment]
note for reviewers: I'm now much happier with the overall direction, but the PR is still WIP.

**Comment by klueska** [comment]
> note for reviewers: I'm now much happier with the overall direction, but the PR is still WIP.

Thanks. Just ping me when it's read for review.

**Comment by fromanirh** [comment]
/hold cancel
@klueska the PR is now ready for review. In general I'd like to add a bit more unit-tests, still thinking about this.

**Comment by klueska** [comment]
I took a quick look at this patch. My first impression is that it seems unnecessarily invasive to plumb all of this `aligned` information through every call while in the process of doing the actual allocation. Can you elaborate on where / how the issue you were seeing manifested and why doing some sort of pre-check isn't sufficient (even if it means doing a more sophisticated pre-check than what we are doing today).

**Comment by bart0sh** [comment]
@fromanirh
> alternative implementation with stricter pre-check here: https://github.com/kubernetes/kubernetes/pull/114114

Should this one be closed in favour of  https://github.com/kubernetes/kubernetes/pull/114114 or you want to keep both PRs open?

**Comment by fromanirh** [comment]
> @fromanirh
> 
> > alternative implementation with stricter pre-check here: #114114
> 
> Should this one be closed in favour of #114114 or you want to keep both PRs open?

Initially I planned to keep both open, but now I'm inclined towards the latter approach. Will reopen this one if we decide the new approach is not good enough (and no other alternatives emerge).

**Pull request closed without merging**